Add documentation page for Managed CodeGen for Native Modules (#5141)
microsoft/react-native-windows: docs/managedCodeGen.md

# Code generation for managed react-native modules

> This is the feature is applicable to C# implemented modules or viewmanagers. It will make sure that the code needed to integrate with the react-native application is constructed at build time instead of at runtime.

Reviewer: ```suggestion
This feature is applicable to C# implemented modules or `ViewManager`s. Its objective is to ensure the code needed to integrate with the `react-native` application is constructed at build time instead of at runtime.
```

> In earlier versions of react-native-windows managed modules registered themselves based on reflection at runtime. This method has a few disadvantages. The error handling was limmited and only availalbe at runtime. The cost for analyzing was paid each and every time the app ran.

Reviewer: ```suggestion
In earlier versions of react-native-windows managed modules registered themselves based on reflection at runtime. This method has a few disadvantages: 
1) the error handling was limited and only available at runtime. 
2) The cost for doing reflection was paid each and every time the app ran. 
3) .NET Native - which is used by UWP apps - has somewhat restrictions on what types/APIs can be reflected upon and complex ways of working around it.
```

Doing it at build time should result in developers seeing all errors up front, rather than only seeing errors for the code being tested. And the runtime cost is not incurred every single time.

## How it works
This change adds a new targets file that is imported from the [CSharpLib.targets](https://github.com/microsoft/react-native-windows/blob/master/vnext/PropertySheets/External/Microsoft.ReactNative.Uwp.CppLib.targets) and [CSharpApp.targets](https://github.com/microsoft/react-native-windows/blob/master/vnext/PropertySheets/External/Microsoft.ReactNative.Uwp.CppApp.targets). That will enable the codegen for managed projects.

### In source mode

> The targets file hooks up an msbuild target that will ensure the tool is deployed and runnable. The codegen tool is defined in the Microsoft.ReactNative.Managed.CodeGen project. It will build and deploy the tool in the current version of your project (except for ARM and Arm64 it will use x86 and x64). It will then collect all the C# source files, assembly references and defines from the app (or library) and invoke the tool to generate a C# file. This is very similar to how Xaml codegen works.

Reviewer: I think the tool itself should always be x64, we don't have to support x86 (and certainly don't have to support ARM and ARM64) **developer** environments.

Author: yea, I changed that in the PR, but missed it here.

> The targets file hooks up an msbuild target that will ensure the tool is deployed and runnable. The codegen tool is defined in the Microsoft.ReactNative.Managed.CodeGen project. It will build and deploy the tool in the current version of your project (except for ARM and Arm64 it will use x86 and x64). It will then collect all the C# source files, assembly references and defines from the app (or library) and invoke the tool to generate a C# file. This is very similar to how Xaml codegen works.

Reviewer: XAML has both codegen and compilation.
The thing you are describing here is similar to XAML compilation (where we generate code-behind for markup files). There is a whole separate infrastructure called XAML codegen that is internal only and is used to produce a bunch of stuff but it's pretty unrelated to this (DXAML, type metadata tables, method tables, mappings between DXAML and Core types).

> The C# file will contain a class called that implements

Reviewer: a class called ??? that implements ??? :)

### In nuget mode (Future)
The tool will be shipped as a binary and the steps won't have to build or deploy the tool. We will also ship one version of the tool, rather than all qualifiers. 

> When building the apps that are checked in to our repo for testing, will remain running off the source versoin.

Reviewer: I'd like our in-repo apps to use the same workflow than we're asking our customers to use

Author: I 100% agree on the point. Unfortunately msbuild and nuget do not support this workflow. I have been a longtime advocate of package is same as source and was one of the driving factors of buildxls dscript design. The only way this would work with msbuild today is if we split the build into multiple solutions that have to be manually restored and build in right order. In fact we will loose F5 for it's unit tests since they have to ensure that the libraries are compiled which puts the unit tests in the second build. I consider this a blocker for our dev workflow.

If you know how to do nuget pack and consume in msbuild in the same solution I would love to learn

> When building the apps that are checked in to our repo for testing, will remain running off the source versoin.

Reviewer: typo: version

## Turning on/off
Currently the feature is behind an msbuild property `$(ReactNativeCodeGenEnabled)`.
The default at the moment is false, it is only turned  on for a single project for now which is the [SampleLibraryCS.csproj](https://github.com/microsoft/react-native-windows/blob/master/packages/microsoft-reactnative-sampleapps/windows/SampleLibraryCS/SampleLibraryCS.csproj) project, to prove it is stable. We will slowly enable it for all projects in the repo and then make it the default.

## MsBuild/Nuget Complications

> MsBuild and Nuget spent a long time fighting me in mixing a NetCoreApp2.2 executable and WinRT apps in the same solution and the same build. ProjectReferences cannot be used so I had to use the `<MsBuild>` task directly in the targets and it was tricky making it build from both the customer apps as well as our main build solution and unittest (Microsoft.ReactNative.sln). In the end there are a few hacks in place to make this work.

Reviewer: We should move to using .net core 5 when it ships (soon). Filed #5146

The hookups in msbuild do not use the latest BeforeTargets/AfterTargets feature to schedule the main work. I opted to follow the same pattern that the Xaml Codegen uses, so the React CodeGen runs after all the other codegen tools (like resgen, xamlcompile etc) to ensure people familiar with those extensions know how to work with the React Managed CodeGen.

## Future improvements
 * Add support for `Task<T>` async methods.

>  * Ship it as a nuget package

Reviewer: are there issues filed for these? can you reference them here?

Author: Yes, will add nrs
